text/0088-tvm-update.md

```diff
@@ -0,0 +1,142 @@
+- **replaces**: -
+- **replaced by**: -
+
+# Summary
+
+This proposal suggests to
+- add new op-codes to TVM related to block data retrieving, hashing, cryptographic signing and message sending
+- include into `c7` register of TVM smart-contract' own code, additional data on transaction and information about masterchain blocks
+ 
+# Motivation
+
+Current TVM implementation makes it hard to implement some common thing. Proposal aims to make it easier and cheaper to:
+- hash big chanks of data (including non-sha256 hashes)
+- work with Bitcoin/Ethereum compatible signtatures
+- prove TON onchain data
+- calculate fee for sending messages
+
+# Guide
+
+Draft implementation is available [here](https://github.com/SpyCheese/ton/commits/new-opcodes).
+
+This proposal suggests to 
+1. Extend **c7** tuple from 10 to 14 elements:
+   - **10**: code of the smart contract.
+   - **11**: value of the incoming message.
+   - **12**: fees collected in the storage phase.
+   - **13**: information about previous blocks.
+2. Add the following TVM op-codes:
+   - **c7** primitives: `MYCODE`, `INCOMINGVALUE`, `STORAGEFEES`, `PREVBLOCKSINFOTUPLE`
+   - Block primitives: `PREVMCBLOCKS`, `PREVKEYBLOCK`
+   - Hash primitives: `HASHSTART`, `HASHEND`, `HASHENDST`, `HASHINFO`, `HASHAPPU`, `HASHAPPI`, `HASHAPPS`, `HASHAPPB`
+   - Cryptography primitives: `ECRECOVER`
+   - Message primitives: `SENDMSG`
+
+3. Change sending fee calculation: deduct fee related to message estimation even in the case of failed sending.
+
+# Specification
+
+## c7 update
+**10th** element of **c7** tuple contains cell with executed smart-contract code (from init_state of incoming message if applicable)
+
+**11th** element of **c7** tuple contains value of incoming message. It is a tuple with two elements: incoming value (same value which is put on stack as 4th element) and a dictionary of extra currencies. If not applicable, the value is `[0, null]`.
+
+**12th** element of **c7** tuple contains fee debited in storage phase.
+
+**13th** elements of **c7** contains of tuple with two elements: `last_mc_blocks` and `prev_key_block`.
+- `last_mc_blocks` is a tuple of up to 16 elements which contains previous most recent masterchain blocks info (ordered by seqno). Each block is represented as tuple of 5 integers: `(workchain, shard, seqno, root_hash, file_hash)`
+- `prev_key_block` is a tuple of 5 integers `(workchain, shard, seqno, root_hash, file_hash)` corresponding to most recent key block.
+
+## TVM update
+### New stack type
+This proposal suggests to introduce new type in TVM `Hasher`: it is opaque type used to consequently hash data chunks. This type behave the same way as others and can be considered immutable: `HASHAPP*` ops (described below) "consume" old hasher and return updated one (that way if old hasher was copied, it's copies are completely independent). This object is not serializable and thus can not be returned from TVM (for instance as result of get-method).
```

Comment:
I see new Hasher object as a not a so good idea. 

IMO ideology of TON is to keep everything as much universal as
possible, e.g everything is a cell, wallets = smart contracts, etc. 

Maybe we can implement other hashes in the same way
as 'HASHCU 'HASHSU works?

I don't actually think that it will be less efficient, I think in the most of use cases we will already have data cell/slice to hash as input.

Reply:
Also, I have an alternative approach which will be same efficent as Hasher, but without introducing new type.

Lets imagine we have opcode `HASH_SHA512` which will accept number of stack elements to hash as first stack element (`s0`).

So if we want to hash 1 uint256 and 1 slice, currently we will do something like:
```
HASHSTART_SHA512
PUSHINT 7777777777
HASHAPPU
PUSHSLICE {001010010101}
HASHAPPS
HASHEND
```
(stack ops are omitted)

But we can do it without hasher type, like:
```
PUSHSLICE {001010010101}
PUSHINT 7777777777
PUSH 2
HASH_SHA512
```
So we read first stack value, it is 2, it means that we need to read 2 more values and hash them, in reverse order. 

This way i think it will be much more clear, and keep architecture cleaner.

Reply:
Not sure which "universality" is broken here.
We are looking for the methods which can be used to hash large chunks of data, like 100kB. Approach with `hasher` objects allows to copy partially hashed data and cheaper get digest of data with the same prefix (not sure that we practicaly need it though).

Reply:
By universality I mean that we introducing new type which cannot be returned and serialized to cell. 

Not sure too, where we can use the feature with copying hasher during single contract call, since it cannot be saved and reused later. 

I could propose 2 options here:
1. Do it without hasher object and without copy feature, HASH_SHA512 approach proposed 1 message before, looks  more efficent and smaller (less stack ops needed).
2. Add cell [de]serialization to hasher, then in could be saved, returned and accepted as input, this way prefix feature has more potential I think. And it is actually not a problem to serialize mid state of hasher and unpack later.

Reply:
> Not sure which "universality" is broken here.
> We are looking for the methods which can be used to hash large chunks of data, like 100kB. Approach with `hasher` objects allows to copy partially hashed data and cheaper get digest of data with the same prefix (not sure that we practically need it though).

how you can hash 100kb data on a blockchain? It requires much more gas that possible.

Reply:
>how you can hash 100kb data on a blockchain? It requires much more gas that possible
Currently yes, but it doesn't look like intended behavior: things which are cheap to nodes (and hashing is cheap) should not be prohibitively expensive for contracts.